synapse/replication/tcp/client.py

```diff
@@ -131,8 +145,66 @@ async def on_rdata(
 
             await self.pusher_pool.on_new_notifications(token, token)
 
+        # Notify any waiting deferreds
+        waiting_list = self._streams_to_waiters.get(stream_name, [])
+
+        # Index of first item with a position after the current token, i.e we
+        # have called all deferreds before this index. If not overwritten by
+        # loop below means either a) no items in list so no-op or b) all items
+        # in list were called and so the list should be cleared. Setting it to
+        # `len(list)` works for both cases.
+        index_of_first_deferred_not_called = len(waiting_list)
+
+        for idx, (position, deferred) in enumerate(waiting_list):
+            if position <= token:
+                try:
+                    with PreserveLoggingContext():
+                        deferred.callback(None)
+                except Exception:
+                    # The deferred has been cancelled or timed out.
+                    pass
+            else:
+                index_of_first_deferred_not_called = idx
+                break
+
+        # (This maintains the order so no need to resort)
+        waiting_list[:] = waiting_list[index_of_first_deferred_not_called:]
```

Comment:
This took me a second to understand. I initially though that deferreds may be dropped without being called, but after running through an example scenario that looks like it's not the case.

Reply:
We're basically just partitioning the list into "deferreds that should be called" and "deferreds that won't called". This drops everything we called in the loop above.

Let me add a comment